lib/client_test.go

```diff
@@ -335,6 +335,88 @@ func TestClient_Query(t *testing.T) {
 	}
 }
 
+func TestClient_Add(t *testing.T) {
+	client, err := firestore.NewClient(context.Background(), "yakiire")
+	if err != nil {
+		panic(err)
+	}
+
+	type fields struct {
+		config    *ClientConfig
+		firestore *firestore.Client
+	}
+	type args struct {
+		ctx        context.Context
+		collection string
+		document   interface{}
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "Adds a doc to the collection",
+			fields: fields{
+				config: &ClientConfig{
+					Credentials: "test",
+					ProjectID:   "yakiire",
+				},
+				firestore: client,
+			},
+			args: args{
+				ctx:        context.Background(),
+				collection: "products",
+				document: map[string]interface{}{
+					"Attributes": map[string]interface{}{
+						"color": "blue", "size": 200,
+					},
+					"CategoryIDs": [...]string{"3", "4", "5"},
+					"ID":          "2",
+					"Name":        "Another Test Product",
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "returns an error when the document is nil",
```

Comment:
Is there a possibility that the input is nil in this case? I don't think so.

Reply:
Since this function is taking an interface unmarshalled from JSON, it will be `nil` if the JSON string is invalid (or empty). I've added an explicit check for this in `cmd/add.go` though, so this might not be necessary.

Reply:
> Since this function is taking an interface unmarshalled from JSON

I think it returns `error` and it should be handled?
Rather than adding the case like that, it seems better to confirm like:

* What should it do in the case if the root level is `Array`?
  * It is the correct JSON format but Firestore won't support it, I guess?

Your idea if not bad but it would be better to think about what kind of input will be put by the user as a parameter 😀

Reply:
> It is the correct JSON format but Firestore won't support it, I guess?
Right. Fixed this by changing it to a `map[string]interface{}` instead, so json unmarshalling will fail for `Array`

> I think it returns error and it should be handled?
Didn't get this point. Should I check for `nil` and make a `New` error in `client.go`?

Reply:
> Didn't get this point. Should I check for nil and make a New error in client.go?
No, I just meant that it would be better to have a possible case and it is also ok not to remove the case anyway.